src/jit/gentree.cpp

```diff
@@ -6586,115 +6803,167 @@ GenTreePtr Compiler::gtNewAssignNode(GenTreePtr dst, GenTreePtr src)
 }
 
 // Creates a new Obj node.
-GenTreeObj* Compiler::gtNewObjNode(CORINFO_CLASS_HANDLE structHnd, GenTree* addr)
+GenTree* Compiler::gtNewObjNode(CORINFO_CLASS_HANDLE structHnd, GenTree* addr)
 {
     var_types nodeType = impNormStructType(structHnd);
     assert(varTypeIsStruct(nodeType));
-    GenTreeObj* objNode = new (this, GT_OBJ) GenTreeObj(nodeType, addr, structHnd);
+    unsigned size = info.compCompHnd->getClassSize(structHnd);
+
+    // It would be convenient to set the GC info at this time, but we don't actually require
+    // it unless this is going to be a destination.
+    if (!varTypeIsStruct(nodeType))
+    {
+        if ((addr->gtOper == GT_ADDR) && (addr->gtGetOp1()->TypeGet() == nodeType))
+        {
+            return addr->gtGetOp1();
+        }
+        else
+        {
+            return gtNewOperNode(GT_IND, nodeType, addr);
+        }
+    }
+    GenTreeBlk* newBlkOrObjNode;
+    {
+        newBlkOrObjNode = new (this, GT_OBJ) GenTreeObj(nodeType, addr, structHnd, size);
+    }
     // An Obj is not a global reference, if it is known to be a local struct.
```

Comment:
Because I failed to eliminate it when I got rid of a condition. Fixed.

---
In reply to: [77258711](https://github.com/dotnet/coreclr/pull/7027#discussion_r77258711) [](ancestors = 77258711)